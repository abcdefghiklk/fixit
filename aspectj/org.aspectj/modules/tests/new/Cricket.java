import org.aspectj.testing.Tester;

public class Cricket {
    public static void main(String[] args) {
        Lib l = new Lib();
        Tester.event("call stringMethod");
        l.stringMethod(2);
        Tester.event("call voidMethod");
        l.voidMethod(2);
        Tester.checkEventsFromFile("Cricket.out");
    }
}


class Lib {
    public void voidMethod(int count) {
        if (count == 0) return;
        else voidMethod(count - 1);
    }

    public String stringMethod(int count) {
        if (count == 0) return "0";
        else return count + "-" + stringMethod(count-1);
    }
}


aspect Trace {
    pointcut entry(): target(Lib) && call(* *(..));
    pointcut topEntry(): entry() && !cflowbelow(entry());

    before(): topEntry() {
        Tester.event("->top entry: " + thisJoinPoint);
    }

    after(): entry() {
        Tester.event("->exit: " + thisJoinPoint);
    }

    after() returning (Object o): entry() {
        Tester.event("->exit: " + thisJoinPoint + " with " + o);
    }
    after(): topEntry() {
        Tester.event("->top exit: " + thisJoinPoint);
    }

    after() returning (Object o): topEntry() {
        Tester.event("->top exit: " + thisJoinPoint + " with " + o);
    }
}
