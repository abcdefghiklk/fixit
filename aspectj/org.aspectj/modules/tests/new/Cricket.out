call stringMethod
->top entry: call(String Lib.stringMethod(int))
->exit: call(String Lib.stringMethod(int))
->exit: call(String Lib.stringMethod(int)) with 0
->exit: call(String Lib.stringMethod(int))
->exit: call(String Lib.stringMethod(int)) with 1-0
->exit: call(String Lib.stringMethod(int))
->exit: call(String Lib.stringMethod(int)) with 2-1-0
->top exit: call(String Lib.stringMethod(int))
->top exit: call(String Lib.stringMethod(int)) with 2-1-0
call voidMethod
->top entry: call(void Lib.voidMethod(int))
->exit: call(void Lib.voidMethod(int))
->exit: call(void Lib.voidMethod(int)) with null
->exit: call(void Lib.voidMethod(int))
->exit: call(void Lib.voidMethod(int)) with null
->exit: call(void Lib.voidMethod(int))
->exit: call(void Lib.voidMethod(int)) with null
->top exit: call(void Lib.voidMethod(int))
->top exit: call(void Lib.voidMethod(int)) with null